tests/chainer_tests/links_tests/loss_tests/test_negative_sampling.py

```diff
@@ -9,77 +9,92 @@
+
+        if self.reduce == 'no':
+            g_shape = self.t.shape
+        elif self.reduce == 'sum':
+            g_shape = ()
+        self.gy = numpy.random.uniform(-1, 1, g_shape).astype(numpy.float32)
 
-    def check_backward(self, x_data, t_data, y_grad):
+    def check_forward(self, x_data, t_data):
         x = chainer.Variable(x_data)
         t = chainer.Variable(t_data)
-        W = self.link.W
+        y = self.link(x, t, reduce=self.reduce)
+        self.assertEqual(y.shape, self.gy.shape)
 
-        y = self.link(x, t)
-        y.grad = y_grad
-        y.backward()
+        W = cuda.to_cpu(self.link.W.data)
+        samples = cuda.to_cpu(y.creator.samples)
 
-        # fix samples
-        negative_sampling.NegativeSamplingFunction.samples = y.creator.samples
+        loss = numpy.empty((len(self.x),), numpy.float32)
+        for i in range(len(self.x)):
+            ix = self.x[i]
+            it = self.t[i]
+            if it == -1:
+                loss[i] = 0
+            else:
+                w = W[samples[i]]
+                f = w.dot(ix)
+                # first one is positive example
+                f[0] *= -1
+                loss[i] = numpy.logaddexp(f, 0).sum()
 
-        def f():
-            return self.link(x, t).data,
-        gx, gW = gradient_check.numerical_grad(
-            f, (x.data, W.data), (y.grad,), eps=1e-2)
-        del negative_sampling.NegativeSamplingFunction.samples  # clean up
+        if self.reduce == 'sum':
+            loss = loss.sum()
 
-        testing.assert_allclose(
-            cuda.to_cpu(gx), cuda.to_cpu(x.grad), atol=1.e-4)
-        testing.assert_allclose(
-            cuda.to_cpu(gW), cuda.to_cpu(W.grad), atol=1.e-4)
-        return x.grad, W.grad
+        testing.assert_allclose(y.data, loss)
+
+    def test_forward_cpu(self):
+        self.check_forward(self.x, self.t)
 
     @attr.gpu
-    @condition.retry(3)
     def test_forward_gpu(self):
-        x = chainer.Variable(self.x)
-        t = chainer.Variable(self.t)
-        y = self.link(x, t)
-
-        self.assertEqual(y.data.dtype, numpy.float32)
-        self.assertEqual(y.data.shape, ())
-
-        # fix samples
-        negative_sampling.NegativeSamplingFunction.samples = cuda.to_gpu(
-            y.creator.samples)
         self.link.to_gpu()
-        y_g = self.link(chainer.Variable(cuda.to_gpu(self.x)),
-                        chainer.Variable(cuda.to_gpu(self.t)))
-        del negative_sampling.NegativeSamplingFunction.samples
+        self.check_forward(cuda.to_gpu(self.x), cuda.to_gpu(self.t))
 
-        self.assertEqual(y_g.data.dtype, numpy.float32)
-        self.assertEqual(y_g.data.shape, ())
+    def check_backward(self, x_data, t_data, w_data, sample, y_grad):
+        t = chainer.Variable(t_data)
+        # `__call__` method of `NegativeSampling` link cannot be tested with
+        # `check_backward` because the link makes different samples on each
```

Comment:
You do use `gradient_check.check_backward` below, which seems contradictory to the comment to me.